sw/device/tests/sim_dv/keymgr_key_derivation.c

```diff
@@ -0,0 +1,220 @@
+    0xa6521d8f, 0x13a0e876, 0x1ca1567b, 0xb4fb0fdf,
+    0x9f89bc56, 0x4bd127c7, 0x322288d8, 0x4e919d54,
+};
+
+/**
+ * Kmac prefix "KMAC" with empty custom string
+ */
+#define KMAC_PREFIX_SIZE 11
+const uint32_t kKmacPrefix[KMAC_PREFIX_SIZE] = {
+    0x4d4b2001, 0x00014341, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+};
+
+const test_config_t kTestConfig;
+
+static void write_info_page(dif_flash_ctrl_state_t *flash, uint32_t page_id,
+                            const uint32_t *data) {
+  uint32_t address = flash_ctrl_testutils_info_region_setup(
+      flash, page_id, kFlashInfoBankId, kFlashInfoPartitionId);
+
+  CHECK(flash_ctrl_testutils_erase_and_write_page(
+            flash, address, kFlashInfoPartitionId, data,
+            kDifFlashCtrlPartitionTypeInfo, kSecretWordSize) == 0);
+
+  uint32_t readback_data[kSecretWordSize];
+  CHECK(flash_ctrl_testutils_read_page(
+            flash, address, kFlashInfoPartitionId, readback_data,
+            kDifFlashCtrlPartitionTypeInfo, kSecretWordSize, 0) == 0);
+  CHECK_BUFFER(data, readback_data, kSecretWordSize);
+}
+
+static void init_flash(void) {
+  dif_flash_ctrl_state_t flash;
+
+  CHECK_DIF_OK(dif_flash_ctrl_init_state(
+      &flash, mmio_region_from_addr(TOP_EARLGREY_FLASH_CTRL_CORE_BASE_ADDR)));
+
+  // Initialize flash secrets.
+  write_info_page(&flash, kFlashInfoPageIdCreatorSecret, kCreatorSecret);
+  write_info_page(&flash, kFlashInfoPageIdOwnerSecret, kOwnerSecret);
+}
+
+/** Place kmac into sideload mode for correct keymgr operation */
+static void init_kmac_for_keymgr(void) {
+  dif_kmac_t kmac;
+  CHECK_DIF_OK(
+      dif_kmac_init(mmio_region_from_addr(TOP_EARLGREY_KMAC_BASE_ADDR), &kmac));
+
+  // Configure KMAC hardware using software entropy.
+  dif_kmac_config_t config = (dif_kmac_config_t){
+      .sideload = true,
+  };
+  CHECK_DIF_OK(dif_kmac_configure(&kmac, config));
+
+  for (uint8_t i = 0; i < KMAC_PREFIX_SIZE; ++i) {
+    mmio_region_write32(mmio_region_from_addr(TOP_EARLGREY_KMAC_BASE_ADDR),
+                        KMAC_PREFIX_0_REG_OFFSET + i * 4, kKmacPrefix[i]);
+  }
+}
+
+void keymgr_advance_state(const dif_keymgr_t *keymgr,
+                          const dif_keymgr_state_params_t *params) {
+  CHECK_DIF_OK(dif_keymgr_advance_state(keymgr, params));
+
+  while (true) {
+    dif_keymgr_status_codes_t status;
+
+    CHECK_DIF_OK(dif_keymgr_get_status_codes(keymgr, &status));
+    if (status == 0) {
+      LOG_INFO("Advancing to next state");
+    } else if (status == kDifKeymgrStatusCodeIdle) {
+      break;
+    } else {
+      LOG_ERROR("Unexpected status: %0x", status);
+      break;
+    }
+  }
+}
+
+void keymgr_check_state(const dif_keymgr_t *keymgr,
+                        const dif_keymgr_state_t exp_state) {
+  dif_keymgr_state_t act_state;
+  CHECK_DIF_OK(dif_keymgr_get_state(keymgr, &act_state));
+  if (act_state != exp_state) {
+    LOG_INFO("Keymgr in unexpected state: %0x, expected to be %0x", act_state,
+             exp_state);
+  }
+}
```

Comment:
These should be testutils as well. They seem reusable in other tests.

Reply:
Agree. Looks like we haven't created a keymgr_testutils. Will create one and add these to it in the next PR.